repair/row_level.cc

```diff
@@ -650,18 +653,59 @@ future<std::list<repair_row>> to_repair_rows_list(repair_rows_on_wire rows, sche
     co_return std::move(row_list);
 }
 
+std::unordered_set<dht::token> get_tokens_from_repair_row_list(const std::list<repair_row>& rows) {
+    std::unordered_set<dht::token> ret;
+    for (auto& row : rows) {
+        // Get the tokens for the rows that are read from the node itself and
+        // ignore the rows received from peer nodes. A row with dirty_on_master
+        // flag set comes from peer node. See to_repair_rows_list.
+        if (!row.dirty_on_master()) {
+            ret.insert(row.get_dk_with_hash()->dk.token());
+        }
+    }
+    return ret;
+}
+
+// Check if the row is expired and it does not shadow any data read from the local node.
+bool can_skip_expired_mf(schema_ptr s, const dht::token& t, mutation_fragment& mf,
+        const std::list<repair_row>& rows,
+        std::optional<std::unordered_set<dht::token>>& keys_from_local_node) {
+    bool skip_write_mf = false;
+    if (mf.is_clustering_row()) {
+        auto& cr = mf.as_clustering_row();
+        auto live = cr.is_live(*s, tombstone(), gc_clock::now());
+        if (!live) {
+            if (!keys_from_local_node) {
+                keys_from_local_node = get_tokens_from_repair_row_list(rows);
+            }
+            bool found_key_in_working_row_buf = keys_from_local_node->contains(t);
+            if (!found_key_in_working_row_buf) {
+                skip_write_mf = true;
```

Comment:
Are you talking about the case a row is not expired because a single cell is not expired but the other 99 cells are expired. We do not optimize it so it is the same as if without the patch in this case.